bridge-token-factory/src/lib.rs

```diff
@@ -372,21 +414,61 @@ impl BridgeTokenFactory {
+            target, message
+        ));
+
         match message {
-            Some(message) => ext_bridge_token::ext(self.get_bridge_token_account_id(token.clone()))
-                .with_static_gas(MINT_GAS)
-                .with_attached_deposit(env::attached_deposit() - required_deposit)
-                .mint(env::current_account_id(), amount.into())
-                .then(
-                    ext_bridge_token::ext(self.get_bridge_token_account_id(token))
-                        .with_static_gas(FT_TRANSFER_CALL_GAS)
-                        .with_attached_deposit(1)
-                        .ft_transfer_call(target, amount.into(), None, message),
-                ),
-            None => ext_bridge_token::ext(self.get_bridge_token_account_id(token))
-                .with_static_gas(MINT_GAS)
-                .with_attached_deposit(env::attached_deposit() - required_deposit)
-                .mint(target, amount.into()),
+            Some(message) => {
+                let mint_self_promise =
+                    ext_bridge_token::ext(self.get_bridge_token_account_id(&token))
+                        .with_static_gas(MINT_GAS)
+                        .with_attached_deposit(self.bridge_token_storage_deposit_required)
+                        .mint(env::current_account_id(), amount.into());
+
+                if amount_to_transfer > 0 {
+                    mint_self_promise.then(
+                        ext_bridge_token::ext(self.get_bridge_token_account_id(&token))
+                            .with_static_gas(FT_TRANSFER_CALL_GAS)
+                            .with_attached_deposit(ONE_YOCTO)
+                            .ft_transfer_call(target, amount_to_transfer.into(), None, message),
+                    )
+                } else {
+                    mint_self_promise
+                }
+            }
+            None => {
+                if fee_amount > 0 {
+                    let mint_fee_promise =
+                        ext_bridge_token::ext(self.get_bridge_token_account_id(&token))
+                            .with_static_gas(MINT_GAS)
+                            .mint(env::current_account_id(), fee_amount.into());
+
+                    if amount_to_transfer > 0 {
+                        ext_bridge_token::ext(self.get_bridge_token_account_id(&token))
+                            .with_static_gas(MINT_GAS)
+                            .with_attached_deposit(self.bridge_token_storage_deposit_required)
+                            .mint(target, amount_to_transfer.into())
+                            .then(mint_fee_promise)
+                    } else {
+                        mint_fee_promise
+                    }
+                } else {
+                    ext_bridge_token::ext(self.get_bridge_token_account_id(&token))
+                        .with_static_gas(MINT_GAS)
+                        .with_attached_deposit(self.bridge_token_storage_deposit_required)
```

Comment:
If the target is already registered, we don't need to attach the deposit here. It will be good to add the possibility don't attach the near tokens for deposit.

Reply:
This improvement is out of the scope of this PR, let's create an issue for that.
Also, note that in addition to this, we need to implement a refund mechanism.